clientconn.go

```diff
@@ -133,12 +140,31 @@ func (cc *ClientConn) resetTransport(closeTransport bool) error {
 		if closeTransport {
 			t.Close()
 		}
-		newTransport, err := transport.NewClientTransport(cc.dopts.protocol, cc.target, cc.dopts.authOptions)
+		// Adjust timeout for the current try.
+		dopts := cc.dopts
+		if dopts.Timeout > 0 {
+			dopts.Timeout -= time.Since(start)
+			if dopts.Timeout <= 0 {
+				cc.Close()
+				return ErrClientConnTimeout
+			}
+		}
+		newTransport, err := transport.NewClientTransport(cc.target, &dopts)
 		if err != nil {
-			// TODO(zhaoq): Record the error with glog.V.
+			if netErr, ok := err.(net.Error); ok && netErr.Timeout() {
+				cc.Close()
+				return ErrClientConnTimeout
+			}
+			sleepTime := backoff(retries)
+			// Fail early before falling into sleep.
+			if dopts.Timeout > 0 && dopts.Timeout < sleepTime + time.Since(start) {
```

Comment:
There was a bug here which is fixed. Your proposal is not working because  transport.NewClientTransport could take some time to return so that at this point the left time to finish dialing is cc.dopts.Timeout - time.Since(start).